radix-engine/src/engine/kernel.rs

```diff
@@ -1374,6 +1407,14 @@ where
 
         Ok(())
     }
+
+    fn emit_application_event(&mut self, event: ApplicationEvent) -> Result<(), RuntimeError> {
+        for m in &mut self.modules {
+            m.on_application_event(&self.current_frame, &mut self.heap, &mut self.track, &event)
+                .map_err(RuntimeError::ModuleError)?;
+        }
+        Ok(())
```

Comment:
As discussed over slack, the should be refactored to:
1. call back `pre_syscall` on modules
2. insert application events to `track.tracked_events`
3. call back `post_syscall` on modules.

Would also consider renaming to `emit_event` to be consistent with `emit_log`, but no blocker.